portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,27 +14,61 @@
 
 package com.google.sps.servlets;
 
+import com.google.gson.Gson;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
-import java.util.ArrayList;
-import com.google.gson.Gson;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  ArrayList<UserComment> userComments = new ArrayList<>();
+
+  /*
+   * Called when a client submits a GET request to the /data URL
+   * Displays all recorded user comments on page
+   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> userComments = new ArrayList<>();
-    userComments.add("Nice website");
-    userComments.add("I like the teal color scheme");
-    userComments.add("Could be improved");
     Gson gson = new Gson();
     String json = gson.toJson(userComments);
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
+
+  /*
+   * Called when a client submits a POST request to the /data URL
+   * Adds submitted comment to internal record if the comment is
+   * non-empty. Then, the page is reloaded.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String userComment = getFieldFromResponse(request, "comment", "");
+    if(userComment.length() != 0) {
+      String userName = getFieldFromResponse(request, "name", "Anonymous");
+      String userEmail = getFieldFromResponse(request, "email", "janedoe@gmail.com");
+      Date date = new Date();
+      String currDate = date.toString();
+      UserComment comment = new UserComment(userName, userEmail, userComment, currDate);
+      userComments.add(comment);
+    }
+    response.sendRedirect("index.html");
+  }
+
+  /*
+   * Extracts the value of fieldName attribute from request object if present
+   * and returns defaultValue if it is not
+   */
+  public String getFieldFromResponse(HttpServletRequest request, String fieldName, String defaultValue) {
+    String fieldValue = request.getParameter(fieldName);
+    if(fieldValue == null) {
```

Comment:
This could be simplified by using Java8 map's `getOrDefault`API.

```
return request.getParameterMap().getOrDefault(fieldName, defaultValue);
```

Reply:
getParameterMap returns an array of strings so converting default value to an array and then extracting the value from the result might be the same amount of work. Is it okay if I keep the current logic?

Reply:
That is a good point.

However, it returns an array of strings for a reason: HTTP requests might contain multiple parameters with the same name. It does not happen in your code, but if somehow you (or someone in the future) made a copy-paste mistake in the HTML, `getParameter` will silently choose the first value, which might cause subtle issues and hard-to-find bugs.

So even if `getParameterMap` does not simplify the code, it might still be preferable for correctness and robustness. Do that makes sense?

Reply:
Makes sense, but in this case wouldn't we just return the first element of the array anyway?

Reply:
You can throw an exception when there are multiple elements in the array.

The "correct approach" is to return an appropriate error response via some kind of exception handler, instead of crashing the server, but you don't need to do that in this PR.

Reply:
Oh I see. I'll do that, thanks a lot!